### PR TITLE
DNN-31366 - Deleted page can still be selected as parent page

### LINKED ISSUE #2919: Deleted page can still be selected as parent page
## Description of bug
There is a bug in the Pages screen when selecting a parent page. 

The deleted page is found inside the Recycle Bin. Behavior changes when the deleted page in Recycle Bin is deleted. Delete page is no longer visible in the Parent Page dropdown.

Note: Only deleted page that has a parent page can still be selected as a parent page. The deleted page that does not have parent page, once deleted is not visible in the Parent Page dropdown.

## Steps to reproduce
1. Create a page called "PAGETODELETE" and select a parent page for it. 
2. After creation, delete PAGETODELETE. 
3. Click Add Page, in the new page form, go to Parent Page dropdown 

## Current result
In Pages screen, you are still able to see PAGETODELETE page.

## Expected result
In Pages screen, you should not be able to see PAGETODELETE page.

## Screenshots
image

## Error log
Paste the error log that is related to this issue.

## Additional context
Add any other context about the bug that may be helpful for resolution.

## Affected version

* [x] 9.3.2 latest supported release
* [x] 9.2.2 latest supported release

## Affected browser

* [x] Chrome
* [x] Firefox
* [x] Safari
* [x] Internet Explorer
* [x] Edge

### PR DESCRIPTION
**Pull request by berkarslan-xo**
Partially Fixes #2919 (There are total 2 PRs; for the other PR, you can check: https://github.com/dnnsoftware/Dnn.AdminExperience/pull/1099 ).

## Summary

A new overload which accepts an additional parameter named as "includeDeletedChildren" is added. (retargeted to "9.4.x" branch, you can reach the previous PR which was targeted to "development" branch from here: https://github.com/dnnsoftware/Dnn.Platform/pull/2920 )

Demo: https://drive.google.com/open?id=1yg6OuVuLMbwCbBlFHXDqQa8Abd3WPIn_

Note: This PR should be merged BEFORE https://github.com/dnnsoftware/Dnn.AdminExperience/pull/1099

### CODE REVIEW

**Comment by berkarslan-xo** [comment]
Dear approvers, any news on this PR?

**Comment by mitchelsellers** [comment]
We are in the middle of the RC process. As soon as the next RC drops we will process.

**Comment by berkarslan-xo** [comment]
CI seems broken (it got a Powershell script error after merging the latest 9.4.x branch) and I don't seem to have the permission to run the build again. Can someone trigger the build please? Btw, I saw some similar failure reasons in some other builds so that there might be an issue with it..

**Comment by mitchelsellers** [comment]
@berkarslan-xo We are working on some issues with the build.  For now I'll trigger a re-build which will at least build clean.

/azp run

**Comment by berkarslan-xo** [comment]
> @berkarslan-xo We are working on some issues with the build. For now I'll trigger a re-build which will at least build clean.
> 
> /azp run

Thanks, the build seems to be not retriggered yet as far as I can see though.

File: DNN Platform/Library/Entities/Tabs/TabController.cs [hunk 1]
```diff
@@ -2580,6 +2622,13 @@ public static void CopyPermissionsToChildren(TabInfo parentTab, TabPermissionCol
                             listTabs.Add(tab);
                         }
                     }
+
+                    // HasChildren should be true in case there is at least one not deleted child
+                    tab.HasChildren &=
+                        includeDeletedChildren || (
```
Comment (Yabgu) [inline DNN Platform/Library/Entities/Tabs/TabController.cs]: what if `GetTabsByParent(tab.TabID, tab.PortalID)` returns empty list, shouldn't `tab.HasChildren` be false at that time? I mean like in the following:

```
tab.HasChildren &= (
        includeDeletedChildren && GetTabsByParent(tab.TabID, tab.PortalID).Any()
    ) || (
        !includeDeletedChildren && GetTabsByParent(tab.TabID, tab.PortalID).Any(a => !a.IsDeleted));
```

or more cleanly:
`GetTabsByParent(tab.TabID, tab.PortalID).Any(a => includeDeletedChildren || !a.IsDeleted));`

Comment (bdukes) [reply DNN Platform/Library/Entities/Tabs/TabController.cs]: If `GetTabsByParent` is empty, I would expect that `tab.HasChildren` is already `false`, so the rest of the conditions aren't considered.  I believe it is fine as it is, but thanks for the review and double check 👍

**Comment by berkarslan-xo** [comment]
Hi @valadas & @bdukes , can one of you give the second approval, please (approvals got reset since a small commit is made)?

**Pull request merged**